src/mapreduce.jl

```diff
@@ -285,3 +285,43 @@ end
         @inbounds return similar_type(a, T, Size($Snew))(tuple($(exprs...)))
     end
 end
+
+struct _InitialValue end
+
+_maybeval(dims::Integer) = Val(Int(dims))
+_maybeval(dims) = dims
+_valof(::Val{D}) where D = D
+
+@inline Base.accumulate(op::F, a::StaticVector; dims = :, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybeval(dims), init)
+
+@inline Base.accumulate(op::F, a::StaticArray; dims, init = _InitialValue()) where {F} =
+    _accumulate(op, a, _maybeval(dims), init)
+
+@inline function _accumulate(op::F, a::StaticArray, dims::Union{Val,Colon}, init) where {F}
+    # Adjoin the initial value to `op`:
+    rf(x, y) = x isa _InitialValue ? y : op(x, y)
+
+    # StaticArrays' `reduce` is `foldl`:
+    results = _reduce(
+        a,
+        dims,
+        (init = (similar_type(a, Union{}, Size(0))(), init),),
+    ) do (ys, acc), x
+        y = rf(acc, x)
+        (vcat(ys, SA[y]), y)
```

Comment:
Actually, we can't use `push` since we need to widen element types as we iterate through the arrays.  Since the function provided by the user may return different types, we need to change the output type appropriately. I added the comment in the code.

Reply:
Oh I see. In that case sure, let's go with what you have.

(Hmm, I wonder whether `push` should have a widening semantic. That may be too weird though.)

Reply:
> I wonder whether `push` should have a widening semantic.

I'd definitely appreciate it :)  I'm using `vcat` from `push!!` to support widening semantics: https://github.com/tkf/BangBang.jl/blob/10a6a7aefe90d4e7948aab977858a3fa34a017c2/src/NoBang/staticarrays.jl#L1